src/passes/GlobalSubtyping.cpp

```diff
@@ -0,0 +1,249 @@
+    }
+
+    // Find and analyze struct operations inside each function.
+    FunctionStructValuesMap<FieldInfo> functionNewInfos(*module),
+      functionSetGetInfos(*module);
+    FieldInfoScanner scanner(functionNewInfos, functionSetGetInfos);
+    scanner.run(runner, module);
+    scanner.runOnModuleCode(runner, module);
+
+    // Combine the data from the functions.
+    StructValuesMap<FieldInfo> combinedNewInfos;
+    StructValuesMap<FieldInfo> combinedSetGetInfos;
+    functionNewInfos.combineInto(combinedNewInfos);
+    functionSetGetInfos.combineInto(combinedSetGetInfos);
+
+    // Propagate things written during new to supertypes, as they must also be
+    // able to contain that type. Propagate things written using set to subtypes
+    // as well, as the reference might be to a supertype if the field is present
+    // there.
+    TypeHierarchyPropagator<FieldInfo> propagator(*module);
+    propagator.propagateToSuperTypes(combinedNewInfos);
+    propagator.propagateToSuperAndSubTypes(combinedSetGetInfos);
+
+    // Combine everything together.
+    combinedNewInfos.combineInto(finalInfos);
+    combinedSetGetInfos.combineInto(finalInfos);
+
+    // While we do the following work, see if we have anything to optimize, so
+    // that we can avoid wasteful work later if not.
+    bool canOptimize = false;
+
+    // We have combined all the information we have about writes to the fields,
+    // but we still need to make sure that the new types makes sense. In
+    // particular, subtyping cares about things like mutability, and we also
+    // need to handle the case where we have no writes to a type but do have
+    // them to subtypes or supertypes; in all these cases, we must preserve
+    // that a field is always a subtype of the parent field. To do so, we go
+    // through all the types downward from supertypes to subtypes, ensuring the
+    // subtypes are suitable.
+    auto& subTypes = propagator.subTypes;
+    UniqueDeferredQueue<HeapType> work;
+    for (auto type : subTypes.types) {
+      if (type.isStruct() && !type.getSuperType()) {
+        work.push(type);
+      }
+    }
+    while (!work.empty()) {
+      auto type = work.pop();
+
+      auto& fields = type.getStruct().fields;
+      for (Index i = 0; i < fields.size(); i++) {
+        auto oldType = fields[i].type;
+        auto& info = finalInfos[type][i];
+        auto newType = info.get();
+        if (newType == Type::unreachable) {
+          // We have no info on this field, so use the old type.
+          info.set(oldType);
+        }
+      }
+
+      if (auto super = type.getSuperType()) {
+        auto& superFields = super->getStruct().fields;
+        for (Index i = 0; i < superFields.size(); i++) {
+          auto newSuperType = finalInfos[*super][i].get();
+          auto& info = finalInfos[type][i];
+          auto newType = info.get();
+          if (!Type::isSubType(newType, newSuperType)) {
```

Comment:
Having stared at this for awhile I'm still having trouble understanding under what circumstances this happens (I think I'm not fully modeling the propagation steps above). Which test case(s) cover this?

Reply:
The testcase beginning on line 422 covers this,

https://github.com/WebAssembly/binaryen/pull/4306/files#diff-f4453daffbce437bfcb3bf0a2117bfdc32fe747707ecfc546f3771c0d34b38e3R422-R466

This is a pretty subtle point and the code should explain it better. I'll push a comment with an example in a minute.